crates/bindings/src/lib.rs

```diff
@@ -45,13 +42,610 @@ pub use timestamp::Timestamp;
+/// ctx.db.{table}().{column}()`
+/// ```
+/// to get a [`UniqueColumn`] from a [`ReducerContext`].
+///
+/// For example:
+/// ```ignore
+/// let by_username: spacetimedb::UniqueColumn<_, String, _> = ctx.db.users().username();
+/// ```
+///
+/// When there is a unique column constraint on the table, insertion can fail if a uniqueness constraint is violated.
+/// If we insert two rows which have the same value of a unique column, the second will fail.
+/// This will be via a panic with [`Table::insert`] or via a `Result::Err` with [`Table::try_insert`].
+///
+/// For example:
+/// ```rust
+/// type CountryCode = String;
+/// type NationalBird = String;
+///
+/// #[table(name = countries)]
+/// struct Country {
+///     #[unique]
+///     code: CountryCode,
+///     national_bird: NationalBird
+/// }
+///
+/// #[reducer]
+/// fn insert_unique_demo(ctx: &ReducerContext, value: u64) {
+///     let result = ctx.db.countries().try_insert(Country {
+///         code: "AU".into(), national_bird: "Emu".into()
+///     });
+///     assert!(result.is_ok());
+///
+///     let result = ctx.db.countries().try_insert(Country {
+///         code: "AU".into(), national_bird: "Great Egret".into()
+///         // Whoops, this was Austria's national bird, not Australia's.
+///         // We should have used the country code "AT", not "AU".
+///     });
+///     // since there's already a country in the database with the code "AU",
+///     // SpacetimeDB gives us an error.
+///     assert!(result.is_err());
+///
+///     // The following line would panic, since we use `insert` rather than `try_insert`.
+///     // let result = ctx.db.countries().insert(Country { code: "CN".into(), national_bird: "Blue Magpie".into() });
+///
+///     // If we wanted to *update* the row for Australia, we can use the `update` method of `UniqueIndex`.
+///     // The following line will succeed:
+///     ctx.db.countries().country_code().update(Country {
+///         code: "AU".into(), national_bird: "Australian Emu".into()
+///     });
+/// }
+/// ```
+///
+/// ### `#[primary_key]`
+///
+/// Implies `#[unique]`. Also generates additional methods client-side for handling updates to the table.
+/// <!-- TODO: link to client-side documentation. -->
+///
+/// ### `#[index(btree)]`
+///
+/// Creates a single-column index with the specified algorithm.
+///
+/// It is an error to specify this attribute together with `#[unique]`.
+/// Unique constraints implicitly create a unique index, which is accessed using the [`UniqueColumn`] struct instead of the
+/// [`BTreeIndex`] struct.
+///
+/// The created index has the same name as the column. <!-- TODO: this may change if we do the unify-index-names PR. -->
+///
+/// For a table *table* and an indexed *column*, use:
+/// ```text
+/// ctx.db.{table}().{column}()
+/// ```
+/// to get a [`BTreeIndex`] from a [`ReducerContext`].
+///
+/// For example:
+///
+/// ```ignore
+/// ctx.db.cities().latitude()
+/// ```
+///    
+/// # Generated code
```

Comment:
I kind of hate this whole section, honestly. Not in the sense that it's poorly written, but in that I don't want to document specifics of the codegen in this way. I'd much rather document how users will interact with the generated code, rather than describing the code itself. The whole point of generating types with horrible ugly names like `country__TableHandle` is that users aren't supposed to name those types. And the whole point of using macros to generate code is that users don't need to know the implementation details of that generated code.

Reply:
My main goal was just to communicate that there is a separate handle type from the row type itself, with a completely different API. Naming the types makes it clear what is going on, even if it is ugly.

Also note that if users look at the rustdoc for the generated code, they will see the TableHandle types.

Reply:
I'm adding a note encouraging users not to name the table handle explicitly.